pkg/controller/logstash/config.go

```diff
@@ -85,3 +116,158 @@ func defaultConfig() *settings.CanonicalConfig {
 
 	return settings.MustCanonicalConfig(settingsMap)
 }
+
+func tlsConfig(useTLS bool) *settings.CanonicalConfig {
+	if !useTLS {
+		return nil
+	}
+	return settings.MustCanonicalConfig(map[string]interface{}{
+		"api.ssl.enabled":           true,
+		"api.ssl.keystore.path":     APIKeystorePath,
+		"api.ssl.keystore.password": APIKeystoreDefaultPass,
+	})
+}
+
+// checkTLSConfig ensures logstash config `api.ssl.enabled` matches the TLS setting of API service
+// we allow disabling TLS in service and leaving `api.ssl.enabled` unset in logstash.yml, otherwise throw error
+func checkTLSConfig(config *configs.APIServer, useTLS bool) error {
+	svcUseTLS := strconv.FormatBool(useTLS)
+	sslEnabled := config.SSLEnabled
+	if (svcUseTLS == sslEnabled) || (!useTLS && sslEnabled == "") {
+		return nil
+	}
+
+	return fmt.Errorf("API Service `spec.services.tls.selfSignedCertificate.disabled` is set to `%t`, but logstash config `api.ssl.enabled` is set to `%s`", !useTLS, sslEnabled)
+}
+
+// resolveAPIServerConfig gives ExpectedAPIServer with the resolved ${VAR} value
+func resolveAPIServerConfig(cfg *settings.CanonicalConfig, params Params) (*configs.APIServer, error) {
+	config := baseAPIServer(cfg)
+
+	if unresolvedConfig := patchWithPatternValue(config); len(unresolvedConfig) > 0 {
+		combinedMap, err := getKeystoreEnvKeyValue(params)
+		if err != nil {
+			return nil, err
+		}
+
+		resolveConfigValue(unresolvedConfig, combinedMap)
+	}
+
+	return config, nil
+}
+
+// baseAPIServer gives api.* configs with the default value
+func baseAPIServer(cfg *settings.CanonicalConfig) *configs.APIServer {
+	enabled, _ := cfg.String("api.ssl.enabled")
+	keystorePassword, _ := cfg.String("api.ssl.keystore.password")
+	authType, _ := cfg.String("api.auth.type")
+	username, _ := cfg.String("api.auth.basic.username")
+	pw, _ := cfg.String("api.auth.basic.password")
+
+	return &configs.APIServer{
+		SSLEnabled:       enabled,
+		KeystorePassword: keystorePassword,
+		AuthType:         authType,
+		Username:         username,
+		Password:         pw,
+	}
+}
+
+// patchWithPatternValue matches the pattern ${VAR:default_value} against config and assign the default value
+// It gives a map of string (VAR) and config pointer that need to further resolve
+// VAR is the variable name that expect to be defined in Keystore or Env
+//
+//	The variable name can consist of digit, underscores and letters
+//	The default value is optional, ${VAR:} and ${VAR} are valid.
```

Comment:
I wonder if we need to support the `default` case, and if we can just stick with the standard `${VAR}`? If we do support the `default` case, we should log that we are using the default

As far as supporting default values, I think that allowing defaults will only work when there are already ENV variables set in the `podTemplate` container definition as it stands, so the use case for supporting defaults is potentially narrow, and potentially a source of confusion. 

Also, I'm not sure if allowing defaults is typical in Kubernetes - I don't believe that `envFrom` supports it's use

Reply:
Supporting default value is mainly for keeping consistent of the feature that Logstash provide. The default value is used when the ${VAR} is missing in env and keystore. It allows Logstash starts without raising error.

```yaml
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: logstash-sample
spec:
  count: 1
  version: 8.11.1
  config:
    api.auth.type: basic
    api.auth.basic.username: "${API_USERNAME:batman}"
    api.auth.basic.password: "${API_PASSWORD:i_am_rich}"
```

Supporting default is not a thing for Kubernetes, but for Stack Monitoring and ReadinessProbe query monitoring endpoint.

Reply:
Yeah, that makes sense, but I'd still like to see a log if we do revert to `default` values (not logging the default values...)

But I'm still on the fence whether we need the complexity of adding support for default values for ENV variables for this feature only, as I suspect adding default values for usernames and passwords is not going to be a massively used feature, and may be something we can document our way around.

Reply:
Agree having default value for username password is a small use case and not a good practice. This makes me think why Logstash supports this feature in `logstash.yml` at the first place. For pipelines config, this is a handy features for business logic. However, for logstash config, all settings are important, so, if I expect a ${VAR} and it is missing, I only want to see error and stop Logstash.

Probably, the default value is just a fallback setup to disable basic authentication and SSL in testing phase.

```yaml
api.ssl.enabled: "${SSL_ENABLED:false}"
api.auth.basic.username: "${API_USERNAME:}"
api.auth.basic.password: "${API_PASSWORD:}"
```

I think it is acceptable to remove the feature. Do you see any downside to keep it? I am open to remove the feature.

Reply:
Yeah, for business logic in pipelines, it's very useful, but it feels a little less so in this case, and silently falling back to defaults for passwords is something I worry about.

How much effort is it to remove? And how much would it simplify?

Reply:
Let's remove it then. I will need to update doc, test cases and the config.go